### PR TITLE
pubsub: allow drivers to specify batching options for ReceiveBatch

### LINKED ISSUE #1725: pubsub: make pubsub call ReceiveBatch concurrently based on batch options
Today, drivers can get a request for `ReceiveBatch` with a large `maxMessages`; if the provider doesn't support fetching that many messages at once, they must make multiple calls concurrently and manage that themselves.

Instead, we could let drivers specify `batcher.Options` and use them to break up the `ReceiveBatch` request based on the min, max batch size, and max concurrency. This would simplify driver code significantly, and also improve retry behavior by retrying each call separately.

This is similar to #1724, but for Receive.

### PR DESCRIPTION
**Pull request by vangent**
This is similar to what I did in #1724, but for `ReceiveBatch`.

I verified that there's no noticeable change in the benchmark times for GCP and AWS.

The `BREAKING_CHANGE_OK` is for drivers, and shouldn't affect users.

Fixes #1725.

### CODE REVIEW

File: pubsub/pubsub.go [hunk 1]
```diff
@@ -641,21 +660,21 @@ var NewSubscription = newSubscription
 
 // newSubscription creates a Subscription from a driver.Subscription.
 //
-// fixedReceiveSize should be 0 except in tests where stability is necessary for
-// record/replay. If non-zero, the maxMessages passed to driver.ReceiveBatch
-// will always be fixedReceiveSize. If 0, maxMessages is determined
-// dynamically based on message throughput.
+// recvBatchOpts sets options for Receive batching. May be nil to accept
+// defaults. The ideal size of messages to receive at a time is determined
```
Comment (jba) [inline pubsub/pubsub.go]: s/size/number/

File: internal/batcher/batcher.go [hunk 2]
```diff
@@ -25,6 +25,52 @@ import (
 	"sync"
 )
 
+// Split splits n into similarly-sized batches based on opts. opts may be nil
+// to accept defaults.
+//
+// Split return nil (e.g., if n is less than o.MinBatchSize).
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/batcher/batcher.go]: Yes, fixed.

File: internal/batcher/batcher.go [hunk 3]
```diff
@@ -25,6 +25,52 @@ import (
 	"sync"
 )
 
+// Split splits n into similarly-sized batches based on opts. opts may be nil
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/batcher/batcher.go]: Yes, it is. Tried again.

File: internal/batcher/batcher.go [hunk 4]
```diff
@@ -25,6 +25,52 @@ import (
...
+		// One batch is fine.
+		return []int{n}
+	}
+	// Compute how many batches we need, including remainder.
+	nBatches := n / o.MaxBatchSize
+	if n%o.MaxBatchSize != 0 {
+		nBatches++
+	}
+	// Given nBatches, see how big each batch should be. Note that we want
+	// nBatches * nBatches to be >= n; if it's > n, the last batch will end up
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/batcher/batcher.go]: Yes, done.

File: internal/batcher/batcher.go [hunk 5]
```diff
@@ -25,6 +25,52 @@ import (
...
+	// being a bit smaller.
+	batchSize := n / nBatches
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/batcher/batcher.go]: You are right. This is surprisingly tricky to do (to support all 3 options). I've replaced with much simpler code that doesn't try to make even-sized batches; it just greedily grabs `MaxBatchSize`. We can try to optimize later.

File: internal/batcher/batcher_test.go [hunk 6]
```diff
@@ -29,6 +29,40 @@ import (
...
+		n    int
+		opts *batcher.Options
+		want []int
+	}{
+		// Defaults.
+		{0, nil, nil},
+		{1, nil, []int{1}},
+		{10, nil, []int{10}},
+		// MinBatchSize.
+		{4, &batcher.Options{MinBatchSize: 5}, nil},
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/batcher/batcher_test.go]: Done.

File: pubsub/gcppubsub/gcppubsub.go [hunk 7]
```diff
@@ -368,6 +335,13 @@ func (s *subscription) pull(ctx context.Context, maxMessages int, returnImmediat
 		}
 		ms = append(ms, m)
 	}
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (vangent) [reply pubsub/gcppubsub/gcppubsub.go]: Why not? It's not new code, it was at line 343 previously.

[the remark this answers is not in the document]
Comment (vangent) [reply pubsub/gcppubsub/gcppubsub.go]: Why not? It's not new, it was on line 343 earlier.

File: pubsub/pubsub.go (hunk 1 above)
Comment (vangent) [reply pubsub/pubsub.go]: Done.

**Comment by vangent** [comment]
PTAL.

File: pubsub/gcppubsub/gcppubsub.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vangent) [reply pubsub/gcppubsub/gcppubsub.go]: Oh, I see; I thought you were referring to the `len(ms) == 0` block. You are right; removed.

[the remark this answers is not in the document]
Comment (vangent) [reply pubsub/gcppubsub/gcppubsub.go]: I moved this condition up earlier to be more clear.

**Pull request merged**